### ISSUE TITLE
Extra padding in Arrow file format?

### ISSUE THREAD

**Issue by chutchinson** [apache/arrow#2559]
![arrow-hex](https://user-images.githubusercontent.com/2136005/45557397-7dd40c80-b80b-11e8-93e5-0f9275401f0e.png)

I am working on an Arrow implementation according to the specification as of the 0.10 release, and I’m confused about what appears to be extra padding in the file format. I generated a record batch and serialized it to the file format using pyarrow (0.10) and inspected the file in a hex editor.

•	The first 6 bytes are the magic (ARROW1) as expected.
•	The next 2 bytes are padding bytes (0x00) up to offset 0x07 (the 8-byte boundary)

As far as I understand, the following bytes should be the streaming format; however, there is another zero byte (padding?) at offset 0x08 (just after the boundary). This byte is followed by a valid message size, and the rest of the format is constructed as expected. Am I missing something in the serialization documentation about padding after the magic? Am I misunderstanding the concept of an 8-byte boundary? 

I am using this documentation as a reference:

https://github.com/apache/arrow/blob/master/format/IPC.md

What is that extra byte doing there? I can't seem to find the definition in the spec.

**Comment by wesm** [comment]
hi @chutchinson do you want to work on your Arrow implementation in the community rather than externally? That would be the ideal way to collaborate and prevent the maintainers here having to go through a bunch of paperwork if you ever want to donate your codebase to the Apache project.

Here's where the magic number and padding is written:

https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/writer.cc#L926

The next bytes written should be the schema:

**Comment by wesm** [comment]
Hm, I am seeing a 9th byte that shouldn't be there:

```
[ RUN      ] TestFileFormat.DictionaryRoundTrip

Breakpoint 1, arrow::ipc::SchemaWriter::WriteSchema (this=0x7fffffff9ff0) at ../src/arrow/ipc/writer.cc:771
771	    RETURN_NOT_OK(internal::WriteMessage(*schema_fb, sink_, &metadata_length));
(gdb) s
std::__shared_ptr<arrow::Buffer, (__gnu_cxx::_Lock_policy)2>::operator* (this=0x7fffffff9b78)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/bits/shared_ptr_base.h:1041
1041		return *_M_ptr;
(gdb) s
arrow::ipc::internal::WriteMessage (message=..., file=0xb670d0, message_length=0x7fffffff99d8) at ../src/arrow/ipc/metadata-internal.cc:961
961	  RETURN_NOT_OK(file->Tell(&start_offset));
(gdb) n
963	  int32_t padded_message_length = static_cast<int32_t>(message.size()) + 4;
(gdb) p start_offset 
$1 = 9
```

Digging in to see what's going on

**Comment by chutchinson** [comment]
I would love to donate the code (C#); however, I don't have freedom over that choice due to my employer. I'll be an advocate for such a decision, but for now I have to keep the implementation under wraps.

**Comment by wesm** [comment]
@chutchinson there's a bug in the C++ library. I'm a bit mystified why the unit tests did not catch it. I'm working on a patch now

**Comment by chutchinson** [comment]
I've put in a formal request to donate our implementation to the Apache project. That's the most I can do at the moment. Is there a way I can reach you @wesm  to help with the process?

**Comment by wesm** [comment]
Yes, please send me an e-mail -- you can find my e-mail address in the commit logs
